### PR TITLE
feat: httproute validation

### LINKED ISSUE #2079: HTTPRoute: Data-Plane Translation
### Problem Statement

https://github.com/Kong/kubernetes-ingress-controller/pull/2081 provided a controller and a starting point for basic parser translate logic for `HTTPRoutes`. The purpose of this task is to finish full spec support for `HTTPRoute`.

### Acceptance

- [x] as a user I can use the `HTTPRoute` API specification with KIC
- [x] if a feature of `HTTPRoute` is not supported by KIC, I find out quick and upfront via API validation and not via logs later
- [x] parser logic for HTTPRoute implemented and covering the basic route matching portion of the spec
- [x] webhook validation for `HTTPRoute` blocking all unimplemented features
- [x] code refactoring and cleanup
- [x] unit tests for refactored parser functions

### PR DESCRIPTION
**Pull request by shaneutt**
**What this PR does / why we need it**:

This adds validation for `HTTPRoute` objects to fail on `CREATE` or `UPDATE`s that would result in an invalid configuration.

**Which issue this PR fixes**

Partially resolves #2079

### CODE REVIEW

File: internal/admission/server.go [hunk 1]
```diff
@@ -287,6 +292,17 @@ func (a RequestHandler) handleValidation(ctx context.Context, request admission.
...
 			return nil, err
 		}
+	case httprouteGVResource:
+		httproute := gatewayv1alpha2.HTTPRoute{}
+		deserializer := codecs.UniversalDeserializer()
+		_, _, err = deserializer.Decode(request.Object.Raw, nil, &httproute)
+		if err != nil {
+			return nil, err
+		}
+		ok, message, err = a.Validator.ValidateHTTPRoute(context.Background(), httproute)
```
Comment (mflendrich) [inline internal/admission/server.go:302]: Can we make a drive-by fix to use the right `context.Context` in this function?

File: internal/admission/validator.go [hunk 2]
```diff
@@ -304,6 +306,59 @@ func (validator KongHTTPValidator) ValidateGateway(
...
+		// we don't try to validate an HTTPRoute if it doesn't have parentRefs
+		// configured, because there's no way we can be sure the route is managed
+		// by this controller implementation.
+		return true, "", nil
+	}
+
+	// in order to be sure whether or not an HTTPRoute resource is managed by this
+	// controller we disallow references to Gateway resources that do not exist.
+	var managedGateways []*gatewayv1alpha2.Gateway
+	for _, parentRef := range httproute.Spec.ParentRefs {
```
Comment (mflendrich) [inline internal/admission/validator.go:322]: duplication of class matching logic defined in https://github.com/Kong/kubernetes-ingress-controller/blob/main/internal/controllers/gateway/gateway_utils.go

File: internal/admission/validator.go [hunk 3]
```diff
@@ -304,6 +306,59 @@ func (validator KongHTTPValidator) ValidateGateway(
...
 
+func (validator KongHTTPValidator) ValidateHTTPRoute(
+	ctx context.Context, httproute gatewayv1alpha2.HTTPRoute,
+) (bool, string, error) {
+	if len(httproute.Spec.ParentRefs) == 0 {
+		// we don't try to validate an HTTPRoute if it doesn't have parentRefs
+		// configured, because there's no way we can be sure the route is managed
+		// by this controller implementation.
+		return true, "", nil
+	}
```
Comment (mflendrich) [inline internal/admission/validator.go:317]: redundant? the loop below will yield the same result.

File: internal/validation/gateway/httproute.go [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+func getParentRefForHTTPRouteGateway(httproute *gatewayv1alpha2.HTTPRoute, gateway *gatewayv1alpha2.Gateway) (parentRef *gatewayv1alpha2.ParentRef, err error) {
+	// search all the parentRefs on the HTTPRoute to find one that matches the Gateway
+	for _, ref := range httproute.Spec.ParentRefs {
+		// determine the namespace for the gateway reference
+		namespace := httproute.Namespace
+		if ref.Namespace != nil {
+			namespace = string(*ref.Namespace)
+		}
+
+		// match the gateway with it's parentRef
```
Comment (mflendrich) [inline internal/validation/gateway/httproute.go:120]:
```suggestion
		// match the gateway with its parentRef
```

File: internal/validation/gateway/httproute.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+	return nil
+}
+
+// -----------------------------------------------------------------------------
+// Validation - HTTPRoute - Private Utility Functions
+// -----------------------------------------------------------------------------
+
+// getParentRefForHTTPRouteGateway produces the parentRef that links to a given Gateway
+// object given the HTTPRoute object which links to it.
+func getParentRefForHTTPRouteGateway(httproute *gatewayv1alpha2.HTTPRoute, gateway *gatewayv1alpha2.Gateway) (parentRef *gatewayv1alpha2.ParentRef, err error) {
```
Comment (mflendrich) [inline internal/validation/gateway/httproute.go:111]: this function may be reusable across future route types if we replace `httproute` with `defaultNamespace string`. This will also simplify the parameter list.

File: internal/validation/gateway/httproute.go [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+		// determine the namespace for the gateway reference
+		namespace := httproute.Namespace
+		if ref.Namespace != nil {
+			namespace = string(*ref.Namespace)
+		}
+
+		// match the gateway with it's parentRef
+		if gateway.Namespace == namespace && gateway.Name == string(ref.Name) {
+			copyRef := ref
+			parentRef = &copyRef
```
Comment (mflendrich) [inline internal/validation/gateway/httproute.go:123]: why is this copying necessary?

Comment (shaneutt) [reply internal/validation/gateway/httproute.go:122]: This avoids taking a pointer to a loop variable which will result in all recipients of the pointer getting the same value (the other values in iteration will be lost).

Trying to take a reference to a loop variable will be caught by lint:

```
G601: Implicit memory aliasing in for loop
```

The alternative solution is to use an index variable in the loop and copy it that way, but I personally have no problem with this way.

File: internal/admission/validator.go (hunk 2 above)
Comment (shaneutt) [reply internal/admission/validator.go:322]: Maybe I'm missing what function you're talking about, or misunderstanding in some other way but I'm not seeing the duplication?

File: internal/admission/server.go (hunk 1 above)
Comment (shaneutt) [reply internal/admission/server.go]: :+1: fixed in 40c231c2d145f61773b6522d1ba5fc39a9e8a9e0

File: internal/admission/validator.go (hunk 3 above)
Comment (shaneutt) [reply internal/admission/validator.go]: sure, changed in 2078fb5a133a95b37654a2bb8891e4a7b966f27f

File: internal/validation/gateway/httproute.go (hunk 4 above)
Comment (shaneutt) [reply internal/validation/gateway/httproute.go]: changed in 2078fb5a133a95b37654a2bb8891e4a7b966f27f

File: internal/admission/validator.go (hunk 2 above)
Comment (mflendrich) [reply internal/admission/validator.go:315]: discussed in zoom: @mflendrich will make a follow up PR fixing the abstraction here

**Review by mflendrich — APPROVED** [review]
